src/coprocessor/codec/convert.rs

```diff
@@ -1370,25 +1359,96 @@ mod tests {
     }
 
     #[test]
-    fn test_bytes_to_f64() {
-        let tests: Vec<(&'static [u8], f64)> = vec![
-            (b"", 0.0),
-            (b" 23", 23.0),
-            (b"-1", -1.0),
-            (b"1.11", 1.11),
-            (b"1.11.00", 0.0),
-            (b"xx", 0.0),
-            (b"0x00", 0.0),
-            (b"11.xx", 0.0),
-            (b"xx.11", 0.0),
+    fn test_convert_bytes_to_f64() {
+        let tests: Vec<(&'static [u8], Option<f64>)> = vec![
+            (b"", None),
+            (b" 23", Some(23.0)),
+            (b"-1", Some(-1.0)),
+            (b"1.11", Some(1.11)),
+            (b"1.11.00", None),
+            (b"xx", None),
+            (b"0x00", None),
+            (b"11.xx", None),
+            (b"xx.11", None),
         ];
 
-        for (v, f) in tests {
-            let ff = super::bytes_to_f64_without_context(v).unwrap();
-            if (ff - f).abs() > EPSILON {
-                panic!("{:?} should be decode to {}, but got {}", v, f, ff);
+        let mut ctx = EvalContext::default();
+        for (i, (v, expect)) in tests.iter().enumerate() {
+            let ff = convert_bytes_to_f64(&mut ctx, v);
+            match expect {
+                Some(val) => {
+                    let got = ff.unwrap();
+                    if (got - val).abs() > EPSILON {
```

Comment:
I guess epsilon is not very necessary. Performing a direct equal compare may be enough.

Reply:
> I guess epsilon is not very necessary. Performing a direct equal compare may be enough.

`0.11` has no precise binary represetation, so i think EPSILON is necessary.

Reply:
I guess the following two results will be identical even if corresponding number is not precise:
- parse it from string
- write it directly as a literal